src/bentoml/_internal/io_descriptors/text_stream.py

```diff
@@ -0,0 +1,163 @@
+from ..utils.http import set_cookies
+from ..utils.lazy_loader import LazyLoader
+from .base import IODescriptor
+
+if t.TYPE_CHECKING:
+    from google.protobuf import wrappers_pb2
+    from typing_extensions import Self
+
+    from ..context import ServiceContext as Context
+    from .base import OpenAPIResponse
+else:
+    wrappers_pb2 = LazyLoader("wrappers_pb2", globals(), "google.protobuf.wrappers_pb2")
+
+MIME_TYPE = "text/event-stream"
+
+
+class TextStream(
+    IODescriptor[t.AsyncIterable[str]], descriptor_id="bentoml.io.TextStream"
+):
+    """
+    :obj:`TextStream` defines API specification for the inputs/outputs of a Service. :obj:`TextStream`
+    represents a stream of string and will be converted from an Asynchronous Generator
+    for all outcoming responses as specified in your API function signature.
+    Input TextStream is not supported for now.
+
+    Currently, stream based output is only supported with a custom Runnable:
+
+    .. code-block:: python
+        :caption: `service.py`
+
+        import bentoml
+        import asyncio
+
+        class StreamRunnable(bentoml.Runnable):
+            SUPPORTED_RESOURCES = ("cpu",)
+            SUPPORTS_CPU_MULTI_THREADING = True
+
+            @bentoml.Runnable.method(stream=True)
+            async def count(self, input_text:str):
+                for i in range(10):
+                    await asyncio.sleep(1)
+                    yield f"{input_text} {i}\n"
+
+        stream_runner = bentoml.Runner(SSERunnable)
+        svc = bentoml.Service("stream", runners=[stream_runner])
+
+        @svc.api(stream=True, input=bentoml.io.Text(), output=bentoml.io.TextStream())
+        async def count(input_text:str):
+            ret = sse_runner.count.async_stream(input_text)
+            return ret
+
+
+    Users then can then serve this service with :code:`bentoml serve`:
+
+    .. code-block:: bash
+
+       % bentoml serve ./service.py:svc --reload
+
+    Users can then send requests to the newly started services with any client:
+
+    .. tab-set::
+
+        .. tab-item:: Bash
+
+           .. code-block:: bash
+
+              % curl -X POST http://0.0.0.0:3000/count --d 'Hello World'
+
+    .. note::
+
+        :obj:`TextStream` is not designed to take any ``args`` or ``kwargs`` during initialization.
+
+    Returns:
+        :obj:`TextStream`: IO Descriptor that represents stream of string type.
+    """
+
+    _mime_type = MIME_TYPE
+
+    def __init__(self, *args: t.Any, **kwargs: t.Any):
+        if args or kwargs:
+            raise BentoMLException(
+                f"'{self.__class__.__name__}' is not designed to take any args or kwargs during initialization."
+            ) from None
+
+    def _from_sample(self, sample: str | bytes):
+        raise NotImplementedError("TextStream does not support input type for now")
+
+    def input_type(self) -> t.Type[str]:
+        raise NotImplementedError("TextStream does not support input type for now")
```

Comment:
We need this for client

```suggestion
    def input_type(self) -> t.Type[str]:
        return str
```

Reply:
If we end up doing this, this should be an async generator, right?

Reply:
oh right this should be async generator